src/main/java/org/openstreetmap/atlas/generator/tools/streaming/resource/HDFSWalker.java

```diff
@@ -27,33 +27,82 @@
  *
  * @author cstaylor
  * @author jklamer
+ * @author sbhalekar
  */
 public final class HDFSWalker
 {
+    /**
+     * Collection of hadoop file status object and it's corresponding depth in the filesystem
+     * hierarchy
+     *
+     * @author sbhalekar
+     */
+    private static final class FileStatusAtDepth
+    {
+        private final int fileDepth;
+        private final FileStatus fileStatus;
+
+        private FileStatusAtDepth(final FileStatus fileStatus, final int fileLevel)
+        {
+            this.fileStatus = fileStatus;
+            this.fileDepth = fileLevel;
+        }
+
+        public int getFileDepth()
+        {
+            return this.fileDepth;
+        }
+
+        public FileStatus getFileStatus()
+        {
+            return this.fileStatus;
+        }
+    }
+
     /**
      * Iterator for BFS over a collection of HDFS files and directories
      *
      * @author cstaylor
      */
     private static final class HDFSIterator extends AbstractIterator<FileStatus>
     {
-        private final Queue<FileStatus> currentPaths;
+
+        /**
+         * This variable represents the depth for all the children of root directory
+         */
+        private static final int ONE = 1;
+
+        private final Queue<FileStatusAtDepth> currentPaths;
 
         private final FileSystem fileSystem;
 
+        /**
+         * Limit to stop traversing the hadoop directory hierarchy. All the file paths would be
+         * explored if this limit exceeds the max depth in filesystem hierarchy. Root will always
+         * have a depth of 0.
+         */
+        private final int maxDepth;
+
         private HDFSIterator(final Path root, final Configuration configuration)
+        {
+            this(root, configuration, HDFSWalker.WALK_ALL);
+        }
+
+        private HDFSIterator(final Path root, final Configuration configuration, final int maxDepth)
         {
             if (root == null)
             {
                 throw new CoreException("Error when creating an HDFSIterator: root can't be null");
             }
+            this.maxDepth = maxDepth;
```

Comment:
Was thinking edge cases, I guess parameter validation doesn't apply here. If we're passed in -100, it would be the same as having no depth.

Reply:
Yes, cases like negative depth or depth values more than the actual depth would be same as having no depth